[Bug]: IsEqualTo overload became ambiguous in 1.40.0 for same-type expected values (CS0121)

### Description

After upgrading from **1.34.5 → 1.40.0**, every `IsEqualTo(...)` call site where the expected value has the same type as the asserted value now fails to compile with **CS0121 — "The call is ambiguous"**.

The regression was introduced by [#5751 — `feat(assertions): IsEqualTo with implicitly-convertible wrappers`](https://github.com/thomhurst/TUnit/pull/5751), which added a second overload alongside the existing one:

```csharp
public static InvokableValueAssertionBuilder<TValue> IsEqualTo<TValue, TOther>(
    this IAssertionSource<TValue> source,
    TOther? expected,
    string? because = null);

public static InvokableValueAssertionBuilder<TValue> IsEqualTo<TValue>(
    this IAssertionSource<TValue> source,
    TValue? expected,
    string? because = null);
```

When `expected` has the same type as `TValue`, both overloads are equally applicable and C# overload resolution can't pick a winner — `TOther = TValue` satisfies the two-generic version exactly as well as the single-generic one binds.

This is a **source-breaking change** affecting effectively every existing test suite that asserts on enums, primitives, value types, records, etc. The 1.40.0 release notes don't flag it as breaking.

`IsEqualTo` ambiguity has been a recurring class of regression — prior history worth referencing:

- [#3872 — IsEqualTo is ambiguous when used `new()`](https://github.com/thomhurst/TUnit/issues/3872) (closed Nov 2025)
- [#3930 — Package upgrade: Ambiguous Member reference](https://github.com/thomhurst/TUnit/issues/3930) (closed Nov 2025)
- [#922 — \[.NET 9\] IsEqualTo is ambiguous for numbers when using `TUnit.Assertions.Extensions.Generic`](https://github.com/thomhurst/TUnit/issues/922) (closed Oct 2024)
- [PR #953 — Improve ambiguity when using `.IsEqualTo(...)`](https://github.com/thomhurst/TUnit/pull/953) (the original ambiguity fix, Oct 2024)

### Expected Behavior

`IsEqualTo(expected)` should compile and bind unambiguously when `expected` is the same type as the value under assertion (the overwhelmingly common case), as it did on 1.39.0 and all earlier versions.

### Actual Behavior

The compiler reports CS0121 and the build fails:

```
error CS0121: The call is ambiguous between the following methods or properties:
  'TUnit.Assertions.Extensions.EqualsAssertionExtensions.IsEqualTo<TValue, TOther>(IAssertionSource<TValue>, TOther?, string?)'
  and
  'TUnit.Assertions.Extensions.EqualsAssertionExtensions.IsEqualTo<TValue>(IAssertionSource<TValue>, TValue?, string?)'
```

### Steps to Reproduce

1. Create a new TUnit test project targeting any current .NET TFM (8/9/10).
2. Reference `TUnit` **1.40.0**.
3. Add a test that asserts equality where the expected value has the same exact type as the asserted value:

   ```csharp
   using System.Net;
   using TUnit.Assertions;
   using TUnit.Assertions.Extensions;

   public class Repro
   {
     [Test]
     public async Task IsEqualTo_Same_Type_Is_Ambiguous()
     {
       using var http = new HttpClient();
       var response = await http.GetAsync("https://example.com/");

       // CS0121 on 1.40.0 — compiled cleanly on 1.39.0 and earlier.
       await Assert.That(response.StatusCode).IsEqualTo(HttpStatusCode.OK);
     }
   }
   ```

4. Run `dotnet build` — observe the CS0121 error.

**Workaround:** specify the generic argument explicitly:

```csharp
await Assert.That(response.StatusCode).IsEqualTo<HttpStatusCode>(HttpStatusCode.OK);
```

### TUnit Version

1.40.0

### .NET Version

.NET 8.0 / .NET 9.0 / .NET 10.0 (reproduces on all three TFMs)

### Operating System

Linux

### IDE / Test Runner

dotnet CLI (dotnet test / dotnet run)

### Error Output / Stack Trace

```shell
error CS0121: The call is ambiguous between the following methods or properties:
  'TUnit.Assertions.Extensions.EqualsAssertionExtensions.IsEqualTo<TValue, TOther>(TUnit.Assertions.Core.IAssertionSource<TValue>, TOther?, string?)'
  and
  'TUnit.Assertions.Extensions.EqualsAssertionExtensions.IsEqualTo<TValue>(TUnit.Assertions.Core.IAssertionSource<TValue>, TValue?, string?)'
```

### Additional Context

- Not using AOT or trimming.
- No custom configuration; stock TUnit setup.
- Reproduces deterministically on every same-type `IsEqualTo` call (independent of whether the type is an enum, primitive, value type, or record).

**Suggested fix** — disambiguate at the API level so callers don't have to:

1. **Constrain the new overload** so it only kicks in when `TOther` is _not_ `TValue` — e.g. by using a distinct method name (`IsEqualToConverted`, `IsEquivalentValue`, ...) or a different parameter shape.
2. **Merge** the two overloads — keep only `IsEqualTo<TValue>(TValue?)` and rely on the language's existing implicit-conversion behaviour, which already binds anything implicitly convertible to `TValue` to that overload.
3. **Disambiguate by receiver type** — make the two-generic overload an extension on a more specialised receiver so overload resolution is decided by the receiver, not by the argument.

Option 2 is the cleanest — the existing `TValue` overload already accepts implicitly-convertible values; the new overload only adds value for cases the language _won't_ implicitly convert, which would be more clearly served by a distinct method name.

### IDE-Specific Issue?

- [x] I've confirmed this issue occurs when running via `dotnet test` or `dotnet run`, not just in my IDE

--- a/TUnit.Assertions.Tests/Bugs/Issue5720Tests.cs
+++ b/TUnit.Assertions.Tests/Bugs/Issue5720Tests.cs
@@ -1,9 +1,4 @@
-// The IsEqualTo<TValue, TOther> overload these tests exercise lives behind
-// `#if NET9_0_OR_GREATER` in TUnit.Assertions (see issue #5765 / #5751) because it
-// relies on [OverloadResolutionPriority] for disambiguation, which is only honored
-// by C# 13+ and only present in System.Runtime.CompilerServices on .NET 9+. The
-// assertions test project multi-targets net8.0/net9.0/net10.0, so this file must
-// match the same gate to keep the net8.0 build green.
+// Gated to match ImplicitConversionEqualityExtensions.cs — see issue #5765.
 #if NET9_0_OR_GREATER
 namespace TUnit.Assertions.Tests.Bugs;
 

--- a/TUnit.Assertions.Tests/Bugs/Issue5720Tests.cs
+++ b/TUnit.Assertions.Tests/Bugs/Issue5720Tests.cs
@@ -1,3 +1,10 @@
+// The IsEqualTo<TValue, TOther> overload these tests exercise lives behind
+// `#if NET9_0_OR_GREATER` in TUnit.Assertions (see issue #5765 / #5751) because it
+// relies on [OverloadResolutionPriority] for disambiguation, which is only honored
+// by C# 13+ and only present in System.Runtime.CompilerServices on .NET 9+. The
+// assertions test project multi-targets net8.0/net9.0/net10.0, so this file must
+// match the same gate to keep the net8.0 build green.
+#if NET9_0_OR_GREATER
 namespace TUnit.Assertions.Tests.Bugs;
 
 /// <summary>
@@ -191,3 +198,5 @@ public class Issue5720Tests
             $"No implicit conversion operator from '{typeof(UnrelatedType)}' to '{typeof(string)}' was found.");
     }
 }
+
+#endif

--- a/TUnit.Assertions.Tests/Bugs/Issue5765Tests.cs
+++ b/TUnit.Assertions.Tests/Bugs/Issue5765Tests.cs
@@ -2,23 +2,10 @@ using System.Net;
 
 namespace TUnit.Assertions.Tests.Bugs;
 
-/// <summary>
-/// Regression tests for GitHub issue #5765:
-/// 1.40.0 introduced a second IsEqualTo overload alongside the source-generated one to
-/// support wrapper Value Objects with implicit conversions (#5720). When called with a
-/// same-type expected value the two overloads were equally applicable and overload
-/// resolution failed with CS0121 — affecting effectively every existing test suite that
-/// compared enums, primitives, value types, or records to a value of their own type.
-///
-/// The implicit-conversion overloads now live behind <c>#if NET9_0_OR_GREATER</c> because
-/// they rely on <c>[OverloadResolutionPriority]</c> to disambiguate, and that attribute is
-/// only honored by C# 13+ (which lines up with .NET 9). On older library targets the new
-/// overload is omitted entirely so the source-generated overload binds without contest.
-/// </summary>
 public class Issue5765Tests
 {
     [Test]
-    public async Task IsEqualTo_SameType_Enum_Compiles_And_Passes()
+    public async Task IsEqualTo_SameType_Enum()
     {
         var status = HttpStatusCode.OK;
 
@@ -26,7 +13,7 @@ public class Issue5765Tests
     }
 
     [Test]
-    public async Task IsEqualTo_SameType_Primitive_Compiles_And_Passes()
+    public async Task IsEqualTo_SameType_Primitive()
     {
         var value = 42;
 
@@ -34,7 +21,7 @@ public class Issue5765Tests
     }
 
     [Test]
-    public async Task IsEqualTo_SameType_Record_Compiles_And_Passes()
+    public async Task IsEqualTo_SameType_Record()
     {
         var point = new Point(1, 2);
 
@@ -42,7 +29,7 @@ public class Issue5765Tests
     }
 
     [Test]
-    public async Task IsNotEqualTo_SameType_Enum_Compiles_And_Passes()
+    public async Task IsNotEqualTo_SameType_Enum()
     {
         var status = HttpStatusCode.OK;
 

--- a/TUnit.Assertions.Tests/Bugs/Issue5765Tests.cs
+++ b/TUnit.Assertions.Tests/Bugs/Issue5765Tests.cs
@@ -1,0 +1,53 @@
+using System.Net;
+
+namespace TUnit.Assertions.Tests.Bugs;
+
+/// <summary>
+/// Regression tests for GitHub issue #5765:
+/// 1.40.0 introduced a second IsEqualTo overload alongside the source-generated one to
+/// support wrapper Value Objects with implicit conversions (#5720). When called with a
+/// same-type expected value the two overloads were equally applicable and overload
+/// resolution failed with CS0121 — affecting effectively every existing test suite that
+/// compared enums, primitives, value types, or records to a value of their own type.
+///
+/// The implicit-conversion overloads now live behind <c>#if NET9_0_OR_GREATER</c> because
+/// they rely on <c>[OverloadResolutionPriority]</c> to disambiguate, and that attribute is
+/// only honored by C# 13+ (which lines up with .NET 9). On older library targets the new
+/// overload is omitted entirely so the source-generated overload binds without contest.
+/// </summary>
+public class Issue5765Tests
+{
+    [Test]
+    public async Task IsEqualTo_SameType_Enum_Compiles_And_Passes()
+    {
+        var status = HttpStatusCode.OK;
+
+        await Assert.That(status).IsEqualTo(HttpStatusCode.OK);
+    }
+
+    [Test]
+    public async Task IsEqualTo_SameType_Primitive_Compiles_And_Passes()
+    {
+        var value = 42;
+
+        await Assert.That(value).IsEqualTo(42);
+    }
+
+    [Test]
+    public async Task IsEqualTo_SameType_Record_Compiles_And_Passes()
+    {
+        var point = new Point(1, 2);
+
+        await Assert.That(point).IsEqualTo(new Point(1, 2));
+    }
+
+    [Test]
+    public async Task IsNotEqualTo_SameType_Enum_Compiles_And_Passes()
+    {
+        var status = HttpStatusCode.OK;
+
+        await Assert.That(status).IsNotEqualTo(HttpStatusCode.NotFound);
+    }
+
+    private sealed record Point(int X, int Y);
+}

--- a/TUnit.Assertions/Extensions/ImplicitConversionEqualityExtensions.cs
+++ b/TUnit.Assertions/Extensions/ImplicitConversionEqualityExtensions.cs
@@ -1,12 +1,7 @@
-// The IsEqualTo<TValue, TOther> / IsNotEqualTo<TValue, TOther> overloads below
-// rely on [OverloadResolutionPriority] to disambiguate against the source-generated
-// IsEqualTo<TValue> / IsNotEqualTo<TValue> overloads when called with a same-type
-// expected value. That attribute is only honored by the C# 13+ compiler and is only
-// present in System.Runtime.CompilerServices on .NET 9+. On older targets the
-// attribute is silently dropped (Polyfill does not supply it), every same-type
-// IsEqualTo call becomes ambiguous (CS0121), and we break effectively every existing
-// test suite. Gating the entire feature to .NET 9+ keeps net8.0 / netstandard2.0
-// consumers on the original well-defined overload. See issue #5765.
+// Gated to .NET 9+ because these overloads rely on [OverloadResolutionPriority] to
+// lose to the source-generated single-generic IsEqualTo / IsNotEqualTo on same-type
+// calls, and that attribute is silently dropped on net8.0 / netstandard2.0 (Polyfill
+// does not supply it), causing CS0121. See issue #5765.
 #if NET9_0_OR_GREATER
 using System.Collections.Concurrent;
 using System.Diagnostics.CodeAnalysis;

--- a/TUnit.Assertions/Extensions/ImplicitConversionEqualityExtensions.cs
+++ b/TUnit.Assertions/Extensions/ImplicitConversionEqualityExtensions.cs
@@ -1,3 +1,13 @@
+// The IsEqualTo<TValue, TOther> / IsNotEqualTo<TValue, TOther> overloads below
+// rely on [OverloadResolutionPriority] to disambiguate against the source-generated
+// IsEqualTo<TValue> / IsNotEqualTo<TValue> overloads when called with a same-type
+// expected value. That attribute is only honored by the C# 13+ compiler and is only
+// present in System.Runtime.CompilerServices on .NET 9+. On older targets the
+// attribute is silently dropped (Polyfill does not supply it), every same-type
+// IsEqualTo call becomes ambiguous (CS0121), and we break effectively every existing
+// test suite. Gating the entire feature to .NET 9+ keeps net8.0 / netstandard2.0
+// consumers on the original well-defined overload. See issue #5765.
+#if NET9_0_OR_GREATER
 using System.Collections.Concurrent;
 using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
@@ -146,3 +156,4 @@ internal static class ImplicitConversionCache
         return null;
     }
 }
+#endif

--- a/TUnit.PublicAPI/Tests.Assertions_Library_Has_No_API_Changes.DotNet8_0.verified.txt
+++ b/TUnit.PublicAPI/Tests.Assertions_Library_Has_No_API_Changes.DotNet8_0.verified.txt
@@ -3767,9 +3767,6 @@ namespace .Extensions
     {
         public static .<TValue> IsEqualTo<TValue>(this .<TValue> source, TValue? expected, [.("expected")] string? expectedExpression = null) { }
         public static .<TValue> IsEqualTo<TValue>(this .<TValue> source, TValue? expected, .<TValue> comparer, [.("expected")] string? expectedExpression = null, [.("comparer")] string? comparerExpression = null) { }
-        [.("Looks up implicit conversion operators via reflection. Trimming may remove user-d" +
-            "efined operators.")]
-        public static .<TOther> IsEqualTo<TValue, TOther>(this .<TValue> source, TOther? expected, [.("expected")] string? expectedExpression = null) { }
     }
     public static class EquatableAssertionExtensions
     {
@@ -4696,9 +4693,6 @@ namespace .Extensions
     public static class NotEqualsAssertionExtensions
     {
         public static .<TValue> IsNotEqualTo<TValue>(this .<TValue> source, TValue notExpected, .<TValue>? comparer = null, [.("notExpected")] string? notExpectedExpression = null, [.("comparer")] string? comparerExpression = null) { }
-        [.("Looks up implicit conversion operators via reflection. Trimming may remove user-d" +
-            "efined operators.")]
-        public static .<TOther> IsNotEqualTo<TValue, TOther>(this .<TValue> source, TOther? notExpected, [.("notExpected")] string? notExpectedExpression = null) { }
     }
     public static class NotEquivalentToAssertionExtensions
     {

--- a/TUnit.PublicAPI/Tests.Assertions_Library_Has_No_API_Changes.Net4_7.verified.txt
+++ b/TUnit.PublicAPI/Tests.Assertions_Library_Has_No_API_Changes.Net4_7.verified.txt
@@ -3360,7 +3360,6 @@ namespace .Extensions
     {
         public static .<TValue> IsEqualTo<TValue>(this .<TValue> source, TValue? expected, [.("expected")] string? expectedExpression = null) { }
         public static .<TValue> IsEqualTo<TValue>(this .<TValue> source, TValue? expected, .<TValue> comparer, [.("expected")] string? expectedExpression = null, [.("comparer")] string? comparerExpression = null) { }
-        public static .<TOther> IsEqualTo<TValue, TOther>(this .<TValue> source, TOther? expected, [.("expected")] string? expectedExpression = null) { }
     }
     public static class EquatableAssertionExtensions
     {
@@ -4189,7 +4188,6 @@ namespace .Extensions
     public static class NotEqualsAssertionExtensions
     {
         public static .<TValue> IsNotEqualTo<TValue>(this .<TValue> source, TValue notExpected, .<TValue>? comparer = null, [.("notExpected")] string? notExpectedExpression = null, [.("comparer")] string? comparerExpression = null) { }
-        public static .<TOther> IsNotEqualTo<TValue, TOther>(this .<TValue> source, TOther? notExpected, [.("notExpected")] string? notExpectedExpression = null) { }
     }
     public static class NotEquivalentToAssertionExtensions
     {

